staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go

```diff
@@ -0,0 +1,196 @@
+				if found {
+					allErrs = append(allErrs, field.Invalid(path, provider, moreThanOneElementErr))
+				} else {
+					allErrs = append(allErrs, validateKeys(provider.AESGCM.Keys, path, aesKeySizes)...)
+					found = true
+				}
+			}
+			if provider.AESCBC != nil {
+				if found {
+					allErrs = append(allErrs, field.Invalid(path, provider, moreThanOneElementErr))
+				} else {
+					allErrs = append(allErrs, validateKeys(provider.AESCBC.Keys, path, aesKeySizes)...)
+					found = true
+				}
+			}
+			if provider.Secretbox != nil {
+				if found {
+					allErrs = append(allErrs, field.Invalid(path, provider, moreThanOneElementErr))
+				} else {
+					allErrs = append(allErrs, validateKeys(provider.Secretbox.Keys, path, secretBoxKeySizes)...)
+					found = true
+				}
+			}
+
+			if provider.Identity != nil {
+				if found {
+					allErrs = append(allErrs, field.Invalid(path, provider, moreThanOneElementErr))
+				} else {
+					found = true
+				}
+			}
+
+			if !found {
+				allErrs = append(allErrs, field.Invalid(path, provider, "provider does not contain any of the expected providers: KMS, AESGCM, AESCBC, Secretbox, Identity"))
+			}
+		}
+	}
+
+	return allErrs
+}
+
+func validateKeys(keys []config.Key, fieldPath *field.Path, expectedLen []int) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(keys) == 0 {
+		allErrs = append(allErrs, field.Required(fieldPath, "at least one key must be provided"))
+		return allErrs
+	}
+
+	for i, key := range keys {
+		allErrs = append(allErrs, validateKey(key, fieldPath.Child("Key").Index(i), expectedLen)...)
+	}
+
+	return allErrs
+}
+
+func validateKey(key config.Key, fieldPath *field.Path, expectedLen []int) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if key.Name == "" {
+		allErrs = append(allErrs, field.Required(fieldPath.Child("Name"), "name is a mandatory field for a keys"))
+	}
+
+	if key.Secret == "" {
+		allErrs = append(allErrs, field.Required(fieldPath.Child("Secret"), "secret is a mandatory field for a keys"))
+		return allErrs
+	}
+
+	secret, err := base64.StdEncoding.DecodeString(key.Secret)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fieldPath.Child("Secret"), "REDACTED", "secret must be base64 encoded"))
+		return allErrs
+	}
+
+	lenMatched := false
+	for _, l := range expectedLen {
+		if len(string(secret)) == l {
+			lenMatched = true
+			break
+		}
+	}
+
+	if !lenMatched {
+		allErrs = append(allErrs, field.Invalid(fieldPath.Child("Secret"), "REDACTED", fmt.Sprintf("secret is not of the expected length: %v", expectedLen)))
+	}
+
+	return allErrs
+}
+
+func validateKMSConfiguration(c *config.KMSConfiguration, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if c.Name == "" {
+		allErrs = append(allErrs, field.Required(fieldPath.Child("Name"), "name is a mandatory field for KMS Providers"))
+	}
+	allErrs = append(allErrs, validateKMSTimeout(c, fieldPath)...)
+	allErrs = append(allErrs, validateKMSEndpoint(c, fieldPath)...)
```

Comment:
I found that it was easier to construct test cases when dealing with tightly scoped validators. 
For example, to validate KMS Timeout value, all I need to send to the method is a KMS config that contains or lacks timeout. In other words, tests a less likely to fail because of the validators that are not under test.

Reply:
But you should also write a top-level test for validateKMSConfiguration
And at that point, it can be stuffed with test cases for inner funcs.

Your tests also check the exact error message, so it's unlikely that a test would fail due to a different check

Reply:
I agree about having to write some top-level tests, but would prefer to do this on a separate PR since this one is getting quite large.